keps/sig-cli/20190318-kustomize-generators-transformers.md

```diff
@@ -0,0 +1,389 @@
+  - This should be possible, but isn't a problem we need to solve right now
+- Develop a market place of any sort
+  - The ecosystem solutions themselves will have market places.  This allows those ecosystem tools to
+    plug their market places into kubectl via Kustomize.
+
+## Proposal
+
+Introduce 2 new types of `exec` plugins:
+
+- Transformers
+- Generators
+
+#### Terms
+
+- *Virtual Resource*
+  - Resource that is not a server-side Kubernetes API object, but used to configure tools
+  - Examples: kubeconfig, kustomization
+- *Generator*
+  - Kustomize directive that accepts a Virtual Resource as input, and emits non-Virtual Resources as output
+  - Examples: `configMapGenerator`, `secretGenerator`
+- *Transformer*
+  - Kustomize directive that accepts 1 Virtual Resource *and* collection of non-Virtual Resources as input
+    and emits non-Virtual Resources as output.
+  - Examples: `commonAnnotations`, `namespace`
+- *Built-In*
+  - Either a Transformer or Generator that is embedded directly in the `kustomization.yaml` rather than
+    as a separate API.
+
+#### Generators
+
+Generators generate 0 or more Resources for some virtual Resource input.
+
+Generator plugins are run immediately after built-in generators:
+
+- `resources`
+- `bases`
+- `secretGenerator`
+- `configMapGenerator`
+
+and before any built-in transformers such as:
+
+- `namePrefix`
+- `commonAnnotations`
+
+Example: Kustomize secretGenerators and configMapGenerators generate Secrets and ConfigMaps
+from a `kustomize.config.k8s.io/v1beta1/Kustomization` virtual Resource.
+
+Generators have 2 components:
+
+1. Generator virtual Resource API definition
+  - By convention, API has a field `generate`
+1. Generator executable plugin that reads the Resources on STDIN and emits generated Resources on STDOUT.
+
+- A Generator is installed by adding the generator executable to
+  `$XDG_CONFIG_HOME/kustomize/plugins/generators/<API group>`.
+- Generators are added as virtual Resources using the `kustomization.yaml` field `generators`
+
+Steps for each `generator`:
+
+1. Kustomize reads the `generator` entry
+1. Kustomize uses the Resource's *Group* to find the plugin under `$XDG_CONFIG_HOME/kustomize/plugins/generators/`
+1. Kustomize execs the plugin it finds, or exits non-0
+1. Kustomize writes the `generator` Resource to the exec process STDIN
+1. Kustomize reads the set of generated Resources from the exec process STDOUT
+1. Kustomize reads error messages from the exec process STDERR
+1. Kustomize fails if the plugin exits non-0
+1. Kustomize adds the emitted Resources to its set of input Resources (e.g. from `resources`, `bases`, etc).
+
+Notes:
+
+- Kustomize will not execute any plugins for the output Resources at this time.
+- If multiple generators are specified, they are run in the order they are listed.
+
+#### Transforms
+
+Transformers modify existing non-virtual Resources by modifying their fields or replacing them with new Resources.
```

Comment:
LMK if this answers any questions:

The `generators` and `transformers` take lists of *Resources*, just like the `resources` field does.

Another way to implement this would be to put what is listed in `generators` and `transformers` under `resources`, have implicitly identify and invoke plugins based on encountering Resources whose group has a transformer or Generator.  I don't really like this approach because it is much hard to debug what is going on in the Kustomization file.

We want to stay as far away as possible from accidentally creating a FSM.

>  should it be in a dryRun mode, where are the chart repos, etc.

We could forward flags to the plugins.

Reply:
> Can Transformers modify/replace/generate virtual Resources too? Why not?

This is something I considered, and don't think we want for v1, but we would consider.

Some thoughts

- We would want to limit this so it wasn't easy for users to create cyclic graphs
- Plugins like `namePrefix` will be hard to make globally idempotent - e.g. re-running them will keep appending.
- Recursing on Virtual Resources isn't necessary within a single Resource Group - since they are handled by the same exec plugin, that plugin can fully expand Virtual Resources internally
- There is dubious need for recursing on Virtual Resources across Resource Groups.

Reply:
Re earlier comment: agreed, I think it's desirable to have kustomisation.yaml only able to describe a fixed transform pipeline.  That allows visualisation and introspection without needing to actually execute the pipeline (at "only" the cost of removing meta-programming).

Fwiw, this means the "virtual resources" don't need to *actually* be resources (they don't need apiVersion/kind/etc structure, they just need to be YAML), since we've just declared they are never acted on as resources.  (I don't expect that to change anything, but I'm just working through the category implications...)

Reply:
> Fwiw, this means the "virtual resources" don't need to actually be resources

Yes this is true.  FWIW, it is the pattern we've follow for both kubeconfig and kustomization.yaml.  It does provide a solution to things we would probably want to tackle - such as the generators detecting if they support a specific version of a Virtual Resource (and failing if they don't), namespacing generators by group, etc.